gitlab/resource_gitlab_repository_file.go

```diff
@@ -0,0 +1,153 @@
+				Required: true,
+			},
+			"branch": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"author_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"author_email": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"commit_message": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabRepositoryFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	file := d.Get("file").(string)
+
+	options := &gitlab.CreateFileOptions{
+		Branch:        gitlab.String(d.Get("branch").(string)),
+		AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+		AuthorName:    gitlab.String(d.Get("author_name").(string)),
+		Content:       gitlab.String(d.Get("content").(string)),
+		CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+		Encoding:      gitlab.String("base64"),
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.CreateFile(project, file, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(repositoryFile.FilePath)
+
+	return resourceGitlabRepositoryFileRead(d, meta)
+}
+
+func resourceGitlabRepositoryFileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	fileID := d.Id()
+	options := &gitlab.GetFileOptions{
+		Ref: gitlab.String(d.Get("branch").(string)),
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.GetFile(project, fileID, options)
+	if err != nil {
+		return err
+	}
+
+	d.Set("project", project)
+	d.Set("file", repositoryFile.FileName)
+	d.Set("content", repositoryFile.Content)
+	d.Set("branch", repositoryFile.Ref)
+
+	return nil
+}
+
+func resourceGitlabRepositoryFileUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	file := d.Get("file").(string)
+	options := &gitlab.UpdateFileOptions{
+		Branch:        gitlab.String(d.Get("branch").(string)),
+		AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+		AuthorName:    gitlab.String(d.Get("author_name").(string)),
+		Content:       gitlab.String(d.Get("content").(string)),
+		CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+		Encoding:      gitlab.String("base64"),
+		//TODO: add LastCommitID
+	}
+
+	if d.HasChange("branch") {
+		options.Branch = gitlab.String(d.Get("branch").(string))
+	}
+
+	if d.HasChange("author_email") {
+		options.AuthorEmail = gitlab.String(d.Get("author_email").(string))
+	}
+
+	if d.HasChange("author_name") {
+		options.AuthorName = gitlab.String(d.Get("author_name").(string))
+	}
+
+	if d.HasChange("content") {
+		options.Content = gitlab.String(d.Get("content").(string))
+	}
+
+	if d.HasChange("commit_message") {
+		options.CommitMessage = gitlab.String(d.Get("commit_message").(string))
+	}
+	_, _, err := client.RepositoryFiles.UpdateFile(project, file, options)
```

Comment:
doesn't work when we change file name

Reply:
I mean it will try to update file which does not exist

Reply:
OK, so whats the exact scenario. Are you updating the `filename` via the `gitlab_repository_file` resource?

Reply:
file resource for the first apply
```hcl
resource "gitlab_repository_file" "this" {
  project        = gitlab_project.this.id
  file           = "XXXXX.tf"
  content        = filebase64("./XXXXX.tf")
  branch         = "master"
  author_name    = "Terraform"
  author_email   = "terraform@example.com"
  commit_message = "feature: add meow file"
}
```

file resource for the second apply:
```hcl
resource "gitlab_repository_file" "this" {
  project        = gitlab_project.this.id
  file           = "YYYYYY.tf"
  content        = filebase64("./XXXXX.tf")
  branch         = "master"
  author_name    = "Terraform"
  author_email   = "terraform@example.com"
  commit_message = "feature: add meow file"
}
```

here I just changed the filename

Reply:
btw, tested file create \ update \ delete, and everything works except the above update

so thanks a lot for your work

Reply:
@mmurygin Awesome! Yeah looks like I didn't add that as an attribute to check for change properly in an update scenario. I wrote some stuff to fix it last night but with the change in TF13 I could not get TF to use my local binary 🤦. I will look into it again tonight and see if I can get that fixed.